core/src/Kokkos_BitManipulation.hpp

```diff
@@ -187,6 +227,35 @@ namespace Kokkos::Impl {
 #define KOKKOS_IMPL_USE_GCC_BUILT_IN_FUNCTIONS
 #endif
 
+template <class T>
+KOKKOS_IMPL_DEVICE_FUNCTION T byteswap_builtin_device(T x) noexcept {
+  return byteswap_fallback(x);
+}
+
+template <class T>
+KOKKOS_IMPL_HOST_FUNCTION T byteswap_builtin_host(T x) noexcept {
+#ifdef KOKKOS_IMPL_USE_GCC_BUILT_IN_FUNCTIONS
+  if constexpr (sizeof(T) == 1) {
+    return x;
+  } else if constexpr (sizeof(T) == 2) {
+    return __builtin_bswap16(x);
+  } else if constexpr (sizeof(T) == 4) {
+    return __builtin_bswap32(x);
+  } else if constexpr (sizeof(T) == 8) {
+    return __builtin_bswap64(x);
+  } else if constexpr (sizeof(T) == 16) {
+#if __has_builtin(__builtin_bswap128)
+    return __builtin_bswap128(x);
+#else
+    return (__builtin_bswap64(x >> 64) |
+            (static_cast<T>(__builtin_bswap64(x)) << 64));
```

Comment:
This is missing the `make_unsigned` intermediary to avoid UB shifting across the sign bit.

Reply:
Then I think we may need to do some ugly `reinterpret_cast<uint64_t>(&x)` business, or really technically, to avoid type punning UB, 
```c++
uint64_t as64[2];
memcpy(as64, &x, 16);
// ... bswap64 on those
```